x-pack/filebeat/input/s3/input.go

```diff
@@ -0,0 +1,449 @@
+		svcSQS := sqs.New(awsConfig)
+		svcS3 := s3.New(awsConfig)
+
+		// receive messages from sqs
+		req := svcSQS.ReceiveMessageRequest(
+			&sqs.ReceiveMessageInput{
+				QueueUrl:              &queueURL,
+				MessageAttributeNames: []string{"All"},
+				MaxNumberOfMessages:   &maxNumberOfMessage,
+				VisibilityTimeout:     &visibilityTimeout,
+				WaitTimeSeconds:       &waitTimeSecond,
+			})
+		output, err := req.Send(context.Background())
+		if err != nil {
+			p.logger.Error("failed to receive message from SQS:", err)
+			continue
+		}
+
+		if len(output.Messages) == 0 {
+			p.logger.Debug("no message received from SQS:", queueURL)
+			continue
+		}
+
+		// process messages received from sqs, get logs from s3 and create events
+		p.processor(queueURL, output.Messages, visibilityTimeout, svcS3, svcSQS)
+	}
+}
+
+// Stop stops the s3 input
+func (p *Input) Stop() {
+	close(p.close)
+	defer p.outlet.Close()
+	p.logger.Info("Stopping s3 input")
+}
+
+// Wait stops the s3 input.
+func (p *Input) Wait() {
+	p.Stop()
+}
+
+func (p *Input) processor(queueURL string, messages []sqs.Message, visibilityTimeout int64, svcS3 *s3.Client, svcSQS *sqs.Client) {
+	var wg sync.WaitGroup
+	numMessages := len(messages)
+	wg.Add(numMessages)
+
+	// process messages received from sqs
+	for i := range messages {
+		errC := make(chan error)
+		// launch goroutine to handle each message from sqs
+		go func(message sqs.Message) {
+			defer wg.Done()
+			defer close(errC)
+
+			s3Infos, err := handleSQSMessage(message)
+			if err != nil {
+				p.logger.Error(errors.Wrap(err, "handelMessage failed"))
+				return
+			}
+
+			// read from s3 object and create event for each log line
+			p.handleS3Objects(svcS3, s3Infos, errC)
+		}(messages[i])
+
+		go func(message sqs.Message) {
+			for {
+				select {
+				case <-p.close:
+					return
+				case err := <-errC:
+					if err != nil {
+						p.logger.Warnf("Processing message failed: %v", err)
+						err := changeVisibilityTimeout(queueURL, visibilityTimeout, svcSQS, message.ReceiptHandle)
+						if err != nil {
+							p.logger.Error(errors.Wrap(err, "change message visibility failed"))
+						}
+						p.logger.Warnf("Message visibility timeout updated to %v", visibilityTimeout)
+					} else {
+						err := deleteMessage(queueURL, *message.ReceiptHandle, svcSQS)
+						if err != nil {
+							p.logger.Error(errors.Wrap(err, "deleteMessages failed"))
+						}
+					}
+					return
+				case <-time.After(time.Duration(visibilityTimeout/2) * time.Second):
+					// If half of the set visibilityTimeout passed and this is
+					// still ongoing, then change visibility timeout.
+					err := changeVisibilityTimeout(queueURL, visibilityTimeout, svcSQS, message.ReceiptHandle)
+					if err != nil {
+						p.logger.Error(errors.Wrap(err, "change message visibility failed"))
+					}
+					p.logger.Infof("Message visibility timeout updated to %v", visibilityTimeout)
+				}
+			}
+		}(messages[i])
+	}
+	wg.Wait()
```

Comment:
NIT: maybe a style comment here to improve visibility, instead of defining the function directly in the body of the processor method I would have define them as real function in the file and just start them as a goroutine:

```golang
go processMessage(message[i], errC)
go sendKeepAlive(message[i], errC)
```

You can unit test them if you want that way.

Reply:
Thanks! I didn't do that because I couldn't name the function 😂